xilinx-2021.07: Replace loading FPGA auth images with the latest patchset

MrCry0 opened the pull request:
This patchset introduces support for the authenticated FPGA images
on ZynqMP boards, besides that introducing common way to pass the
compatible property to any fpga driver.
  
It bases on the initial work by Jorge Ramirez-Ortiz <jorge@founries.io>
  https://patchwork.ozlabs.org/project/uboot/patch/20211015091506.2602-1-jorge@foundries.io/
  https://patchwork.ozlabs.org/project/uboot/patch/20211005111324.19749-3-jorge@foundries.io/
  
**Changes in v5:**

- replace ifdef with if() where it's possible
  
**Changes in v4:**  

- change interface to xilinx_desc->operations->open() callback.
- fix a bug from previous version of the patchset in dereferencing of a parent fpga_desc structure.
  
**Changes in v3:**

- remove the patch which introduced CMD_SPL_FPGA_LOAD_SECURE.
- fix mixing definitions/declarations.
- replace strcmp() calls with more secure strncmp().
- document the "u-boot,zynqmp-fpga-ddrauth" compatible string.
- fix code style by check-patch recommendations.
  
**Changes in v2:**

- add function fit_fpga_load() to simplify calls of fpga_load() from contexts without a compatible attribute.
- move all ZynqMP-specific logic to drivers/fpga/zynqmppl.c
- prepare for passing a "compatible" FDT property to any fpga driver.

MrCry0 committed:
[FIO fromlist] fpga: add option for loading FPGA secure bitstreams

It allows using this feature without enabling the "fpga loads"
command.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Tested-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
[FIO fromlist] fpga: add fit_fpga_load function

Introduce a function which passes an fpga compatible string from
FIT images to FPGA drivers. This lets the different implementations
decide how to handle it.

Some code of Jorge Ramirez-Ortiz <jorge@foundries.io> is reused.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Tested-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
[FIO fromlist] fpga: zynqmp: support loading authenticated images

Add supporting new compatible string "u-boot,zynqmp-fpga-ddrauth" to
handle loading authenticated images (DDR).

Based on solution by Jorge Ramirez-Ortiz <jorge@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Co-developed-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
Tested-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
Revert "[FIO toup] fpga: zynqmp: support loading authenticated images"

The better patchset was upstreamed and applied. Replace the current
implementation of loading fpga authenticated images with upstreamed
one.

This reverts commit d715c11c41b6bbd81b1042a6e23dd621ea963710.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
Revert "[FIO toup] fpga: xilinx: store compatible attribute"

The better patchset was upstreamed and applied. Replace the current
implementation of loading fpga authenticated images with upstreamed
one.

This reverts commit c70d6c95d598b000c302069536927e93f500b6b4.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
Revert "[FIO fromlist] fpga_load: pass compatible string"

The better patchset was upstreamed and applied. Replace the current
implementation of loading fpga authenticated images with upstreamed
one.

This reverts commit e758af5be2c3b587311e0557851802cde5ce4655.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
Revert "[FIO toup] cmd: fpga: Separating the support of fpga loads"

The better patchset was upstreamed and applied. Replace the current
implementation of loading fpga authenticated images with upstreamed
one.

This reverts commit f9871ffe2cabecac49a9a0b04342f65a61d3aad2.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
Revert "[FIO toup] fpga: add option for loading FPGA secure bitstreams"

The better patchset was upstreamed and applied. Replace the current
implementation of loading fpga authenticated images with upstreamed
one.

This reverts commit 06576594aee6da7e4740402e77b02fc1a0290209.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

ricardosalveti commented:
Unfortunately this ended up breaking the secure load process, and it seems to be caused by the difference at how the compatible string is now used (via container_of instead of via function argument).

```
        if ((devnum >= 0) && (devnum < next_desc)) {
                desc = &desc_table[devnum];
                printf("%s: found fpga descriptor #%d @ 0x%p\n",
                      __func__, devnum, desc);
        }
```
` fpga_get_desc: found fpga descriptor #0 @ 0xc0
`

Testing the offset by doing a similar container_of at zynqmp_load:

```
zynqmp_load: desc 0xfffe25b8
zynqmp_load: fdesc 0xfffe25b0
```

The valid fdesc is available at 0xc0 and not 0xfffe25b0.

MrCry0 commented:
> Unfortunately this ended up breaking the secure load process, and it seems to be caused by the difference at how the compatible string is now used (via container_of instead of via function argument).
> 
> ```
>         if ((devnum >= 0) && (devnum < next_desc)) {
>                 desc = &desc_table[devnum];
>                 printf("%s: found fpga descriptor #%d @ 0x%p\n",
>                       __func__, devnum, desc);
>         }
> ```
> 
> `fpga_get_desc: found fpga descriptor #0 @ 0xc0`
> 
> Testing the offset by doing a similar container_of at zynqmp_load:
> 
> ```
> zynqmp_load: desc 0xfffe25b8
> zynqmp_load: fdesc 0xfffe25b0
> ```
> 
> The valid fdesc is available at 0xc0 and not 0xfffe25b0.

Thanks, @ricardosalveti! I'll look at it carefully tomorrow.

MrCry0 commented:
> `fpga_get_desc: found fpga descriptor #0 @ 0xc0`
> Testing the offset by doing a similar container_of at zynqmp_load:

Fixed in the same PR.

MrCry0 committed:
[FIO fromlist] fpga: xilinx: pass an address of xilinx_desc in fpga_desc

Pass an address of xilinx_desc pointer in an fpga_desc  to use parent
fpga_desc structure members inside a xilinx fpga driver.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Tested-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

ricardosalveti commented:
Yup, this now works fine.

Send it upstream and let's wait for the feedback, we can update this pr with whatever lands upstream as we're not blocked since the current implementation is working fine as well.

MrCry0 commented:
> Yup, this now works fine.
> 
> Send it upstream and let's wait for the feedback, we can update this pr with whatever lands upstream as we're not blocked since the current implementation is working fine as well.

Need one more commit - to make this change of xilinx_ops interface general, for all drivers used the interface.

MrCry0 committed:
[FIO fromlist] fpga: xilinx: pass xilinx_desc pointer address into load() ops

Pass an address of xilinx_desc pointer in an fpga_desc into a load()
callback of struct xilinx_fpga_op.
It allows getting parent fpga_desc structure members inside xilinx
fpga drivers.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Tested-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 commented:
> Yup, this now works fine.
> 
> Send it upstream and let's wait for the feedback, we can update this pr with whatever lands upstream as we're not blocked since the current implementation is working fine as well.

@ricardosalveti Sorry, but I had to add one extra commit - to fix stuff for other xilinx fpga drivers :)
If you want, we can retest this final version. Otherwise, I'll upstream it asap. But with RFC :( - we change the common xilinx interface :(

MrCry0 committed:
[FIO fromlist] fpga: xilinx: add missed identifier names

Function definition arguments should also have identifier names.
Add missed ones to struct xilinx_fpga_op callbacks, unifying code.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

Tim-Anderson reviewed (approved):
LGTM
